src/librustc_mir/lints.rs

```diff
@@ -0,0 +1,130 @@
+use rustc_data_structures::bit_set::BitSet;
+use rustc::hir::def::Def;
+use rustc::hir::def_id::DefId;
+use rustc::hir::intravisit::FnKind;
+use rustc::lint::builtin::UNCONDITIONAL_RECURSION;
+use rustc::mir::{self, Mir, TerminatorKind};
+use rustc::ty::{Instance, TyCtxt, TyKind};
+use rustc::ty::subst::Substs;
+
+pub fn check_fn_for_unconditional_recursion(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                            fn_kind: FnKind,
+                                            mir: &Mir<'tcx>,
+                                            def_id: DefId) {
+    if let FnKind::Closure(_) = fn_kind {
+        // closures can't recur, so they don't matter.
+        return;
+    }
+
+    // Walk through this function (say `f`) looking to see if
+    // every possible path references itself, i.e. the function is
+    // called recursively unconditionally. This is done by trying
+    // to find a path from the entry node to the exit node that
+    // *doesn't* call `f` by traversing from the entry while
+    // pretending that calls of `f` are sinks (i.e. ignoring any
+    // exit edges from them).
+    //
+    // NB. this has an edge case with non-returning statements,
+    // like `loop {}` or `panic!()`: control flow never reaches
+    // the exit node through these, so one can have a function
+    // that never actually calls itself but is still picked up by
+    // this lint:
+    //
+    //     fn f(cond: bool) {
+    //         if !cond { panic!() } // could come from `assert!(cond)`
+    //         f(false)
+    //     }
+    //
+    // In general, functions of that form may be able to call
+    // itself a finite number of times and then diverge. The lint
+    // considers this to be an error for two reasons, (a) it is
+    // easier to implement, and (b) it seems rare to actually want
+    // to have behaviour like the above, rather than
+    // e.g. accidentally recursing after an assert.
+
+    let basic_blocks = mir.basic_blocks();
+    let mut work_queue = vec![mir::START_BLOCK];
```

Comment:
Can you rename this to `reachable_without_self_call_queue` or some shorter variant of it?